File: LeetCodeKotlin/KthSmallestElementInBST_230.kt
class KthSmallestElementInBST_230 {
    fun kthSmallest(root: TreeNode?, k: Int): Int {
        // We'll do an iterative inorder traversal
        val stack = ArrayDeque<TreeNode>()
        var current = root
        var count = 0

        while (true) {
            // 1) Push all the left children onto the stack
            while (current != null) {
                stack.addLast(current)
                current = current.left
            }

            // 2) Pop from stack, this is the next smallest node
            val node = stack.removeLast()
            count++
            if (count == k) {
                // The k-th node visited is our result
                return node.value
            }

            // 3) Move to the right subtree of the popped node
            current = node.right
        }
    }
}

File: LeetCodeCpp/problems/trees/KthSmallestElementInBST_230.cpp
#include "KthSmallestElementInBST_230.h"

#include <stack>

using namespace std;

int KthSmallestElementInBST_230::kthSmallest(TreeNode<int>* root, int k) {
    stack<TreeNode<int>*> nodes;
    TreeNode<int>* node = root;

    while (node != nullptr || !nodes.empty()) {
        while (node != nullptr) {
            nodes.push(node);
            node = node->left;
        }

        node = nodes.top();
        nodes.pop();

        if (--k == 0) {
            return node->val;
        }

        node = node->right;
    }

    return -1;
}
